### DISCUSSION TITLE
Hiding standard library definitions?

### DISCUSSION THREAD

**Discussion by osa1** [koka-lang/koka#621] [Q&A]
I have my own `println` that uses my own `console` effect that I can implement handlers for.

When I use it I get errors like

```
guess-the-number.kk(50, 7): type error: identifier println cannot be resolved
  context      :       println("That's not an int!")
  inferred type: (string) -> _
  candidates   : println
                 console/string/println
                 console/show/println(_,show/string/show)
  hint         : qualify the name?

Failed to compile guess-the-number.kk
```

(Note: I think the first candidate should be shown as `current-module/println`, currently it's unclear for a beginner how to qualify it)

Since I will be using my own println in this module many times, it would make sense for me to hide the other definitions so that a standalone `println` would mean the one defined in the current module, and I could refer to the others with the fully qualified name like `core/console/println`.

Is there a way to do this?

**Comment by TimWhiting** [comment]
Not currently, but I also think there should be a way of hiding. For now just use the qualified name (you are correct that the first candidate could be referred to using the current-module, and I agree that we should probably show the fully qualified name in parentheses after the unqualified name for the first candidate here.

**Comment by osa1** [comment]
FWIW, this also makes reading the papers more difficult. For example I can't copy the code in "Structured Asynchrony with Algebraic Effects" as they override some of the standard library definitions.

**Reply by TimWhiting** [reply]
The papers have the added disadvantage that they were written assuming std library functions from previous versions as well. So not only will there be std definitions that conflict, there might be some that are missing as well. That paper in particular was written for v1.x.x koka I believe whereas now we are on v3.x.x. I believe v1.x.x only had a Javascript / C# backend. When moving to the C backend as the primary backend some things were not immediately available (an event loop), which is why we don't have async reintroduced.

**Reply by TimWhiting** [reply]
We should probably keep up-to-date versions of the paper code for the current version, with notes on what changes were necessary or why something might not be possible or needs larger refactoring (maybe samples/papers/...?). I was not a contributor at that point in time though so I'd be working through it the same as you.

**Reply by osa1** [reply]
> We should probably keep up-to-date versions of the paper code for the current version

I think this is a good idea. I'm currently trying to reproduce the paper "Structured Asynhcrony with Algebraic Effects", I can send a PR once it's working, and maybe others can contribute code from other papers.
